File: SimpleEngine/fse/Binder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SimpleEngine.fse
{
    public class Binder
    {
        Map<Shader> shaders;

        public Binder(FileSystem sys)
        {
            shaders = new Map<Shader>(delegate(string name) { return new Shader(sys, name); });
        }

        internal Shader getShader(string shadername)
        {
            return shaders.get(shadername);
        }

        internal Shader getShaderOrNull(string shadername)
        {
            if (string.IsNullOrEmpty(shadername)) return null;
            return shaders.get(shadername);
        }

        List<BufferReference> references = new List<BufferReference>();

        internal void reference(BufferReference br)
        {
            if (references.Contains(br))
            {
                throw new Exception(br.Name + " already addd to " + this.ToString());
            }
            references.Add(br);
        }

        class FboCreator
        {
            Pool<Fbo> pool;
            Map<Fbo> map;

            public FboCreator(int width, int height)
            {
                pool = new Pool<Fbo>(delegate() { return new Fbo(width, height, false); });
                map = new Map<Fbo>(delegate(string n) { return pool.get(); });
            }

            public Fbo create(string name)
            {
                return map.get(name);
            }

            public void release(Fbo fbo)
            {
                pool.release(fbo);
            }
        }

        Dictionary<Size, FboCreator> creators = new Dictionary<Size, FboCreator>( new Size.Comperator() );

        //Pool<Fbo> cbs = new Pool<Fbo>(delegate() { return new Fbo(512, 512, false); });
        //Map<Fbo> buffers = new Map<Fbo>(delegate(string n) { return allocate(n); });
        private Fbo allocate(string name)
        {
            return getCreator(name).create(name);
        }

        private FboCreator getCreator(string name)
        {
            FboCreator c;
            Size size = sizeOf(name);
            if (false == creators.ContainsKey(size))
            {
                c = new FboCreator(size.Width, size.Height);
                creators.Add(size, c);
            }
            else
            {
                c = creators[size];
            }
            return c;
        }
        private void release(Fbo buff)
        {
            creators[new Size { Width=buff.Width, Height=buff.Height }].release(buff);
        }

        private Size sizeOf(string name)
        {
            if (false == associations.ContainsKey(name))
            {
                throw new Exception(name + " is missing a defined size");
            }
            else return associations[name];
        }

        internal void createBuffers()
        {
            StringCounter sc = new StringCounter();
            foreach (BufferReference b in references)
            {
                sc.add(b.Name);
            }

            StringCounter usages = new StringCounter();
            foreach (BufferReference b in references)
            {
                Fbo buff = allocate(b.Name);
                b.setBuffer(buff);
                usages.add(b.Name);
                if (usages.countsOf(b.Name) == sc.countsOf(b.Name))
                {
                    release(buff);
                }
            }
        }

        Dictionary<string, Size> associations = new Dictionary<string, Size>();
        internal void associate(string p, Size size)
        {
            associations.Add(p, size);
        }
    }
}

File: src/engine/fse/binder.cc
#include "engine/fse/binder.h"

#include <algorithm>

#include "engine/stringcounter.h"

namespace simple::fse
{
    FboCreator::FboCreator(int width, int height)
        : pool([width, height]() { return std::make_shared<Fbo>(width, height, false); })
        , map([this](const std::string& n) { return pool.Create(); })
    {
    }

    std::shared_ptr<Fbo> FboCreator::create(const std::string& name)
    {
        return map.GetOrCreate(name);
    }

    void FboCreator::Destroy(std::shared_ptr<Fbo> fbo)
    {
        pool.Destroy(fbo);
    }

    std::string Binder::ToString() const
    {
        return "<unknown binder>";
    }

    Binder::Binder(std::shared_ptr<FileSystem> sys)
        : shaders([sys](const std::string& name) { return std::make_shared<ShaderProgram>(sys.get(), name); })
    {
    }

    std::shared_ptr<ShaderProgram> Binder::GetShader(const std::string& shadername)
    {
        return shaders.GetOrCreate(shadername);
    }

    std::shared_ptr<ShaderProgram> Binder::GetShaderOrNull(const std::string& shadername)
    {
        if (shadername.empty())
        {
            return nullptr;
        }
        return shaders.GetOrCreate(shadername);
    }

    void Binder::AddReference(std::shared_ptr<BufferReference> br)
    {
        if (std::find(references.begin(), references.end(), br) != references.end())
        {
            throw std::runtime_error(fmt::format("{} already addd to {}", br->name, ToString()));
        }
        references.emplace_back(br);
    }

    std::shared_ptr<Fbo> Binder::CreateFbo(const std::string& name)
    {
        return GetCreator(name)->create(name);
    }

    std::shared_ptr<FboCreator> Binder::GetCreator(const std::string& name)
    {
        Size size = GetSize(name);
        auto found = creators.find(size);
        if (found == creators.end())
        {
            auto c = std::make_shared<FboCreator>(size.Width, size.Height);
            creators.emplace(size, c);
            return c;
        }
        else
        {
            return found->second;
        }
    }

    void Binder::DestroyFbo(std::shared_ptr<Fbo> buff)
    {
        const auto size = Size{buff->Width(), buff->Height()};
        auto found = creators.find(size);
        if (found != creators.end())
        {
            found->second->Destroy(buff);
        }
    }

    Size Binder::GetSize(const std::string& name)
    {
        auto found = associations.find(name);
        if (found == associations.end())
        {
            throw std::runtime_error(fmt::format("{} is missing a defined size", name));
        }
        else
        {
            return found->second;
        }
    }

    void Binder::CreateBuffers()
    {
        StringCounter sc;
        for (auto b : references)
        {
            sc.Add(b->name);
        }

        StringCounter usages = StringCounter();
        for (auto b : references)
        {
            auto buff = CreateFbo(b->name);
            b->SetBuffer(buff);
            usages.Add(b->name);
            if (usages.GetCountsOf(b->name) == sc.GetCountsOf(b->name))
            {
                DestroyFbo(buff);
            }
        }
    }

    void Binder::SetSize(std::string p, Size size)
    {
        associations.emplace(p, size);
    }
}
